### PR TITLE
Add more support for json testing scenarios

### PR DESCRIPTION
**Pull request by nwatson22**
This PR adds support for app call and asset config transactions, and assets in the setup-network stage. It also makes some changes to the way panics are handled to make the system compatible with the new system for checking the expected return codes and panic codes. Since the error code checking is done inside the k execution, this also allows us to run our tests faster since `kore-print` can be called with the `--output none` option, and printing with `--output json` was taking up the bulk of the time before in running the json test scenarios.

### CODE REVIEW

File: kavm/src/kavm/__main__.py [hunk 1]
```diff
@@ -136,12 +139,14 @@ def exec_run(
                 print(proc_result.stdout)
             exit(proc_result.returncode)
         elif input_file.suffix == '.json':
+            print("test1")
```
Comment (geo2a) [inline kavm/src/kavm/__main__.py:142]:
```suggestion
```

File: kavm/src/kavm/kavm.py [hunk 2]
```diff
@@ -166,19 +167,48 @@ def run_avm_json(
...
+            print(f'Test file {input_file} does not contain an "execute-transactions" stage')
+            exit(1)
+        for txn in execute_transactions_stage['data']['transactions']:
+            if 'apap' in txn:
+                teal_paths.add(txn['apap'])
+            if 'apsu' in txn:
+                teal_paths.add(txn['apsu'])
+
+        def run_process_on_bison_parser(path: Path) -> str:
+            print(teal_programs_parser)
```
Comment (geo2a) [inline kavm/src/kavm/kavm.py:184]: This `print` must either be dropped or refactored to use the logger.

File: kavm/src/kavm/kavm.py [hunk 3]
```diff
@@ -166,19 +167,48 @@ def run_avm_json(
...
+            )
+            teal_kore_map_item = map_item_op + '(' + teal_path_parsed + ',' + teal_parsed + ')'
+            current_teal_pgms_map = map_union_op + "(" + current_teal_pgms_map + "," + teal_kore_map_item + ")"
 
         krun_command = ['krun', '--definition', str(self.definition_dir)]
         krun_command += ['--output', output]
-        krun_command += [f'-cTEAL_PROGRAMS={teal_programs}']
-        krun_command += [f'-pTEAL_PROGRAMS={str(teal_programs_parser)}']
+        krun_command += [f'-cTEAL_PROGRAMS={current_teal_pgms_map}']
+        krun_command += ['-pTEAL_PROGRAMS=cat']
```
Comment (geo2a) [inline kavm/src/kavm/kavm.py:205]: I am confused by the use of `cat` here. If I recall correctly, we have agreed to use the generated parser for `TealProgamsStore` to parse to parse the bundled Teal programs. In fact, after looking at this PR, I agree that we can as well drop the extra `TealProgamsStore` sort and use `Map` with `Id` keys and `TealInputPgm` values (as you are doing now).

We should drop the now unused definition of `TealProgamsStore` from `avm/teal/teal-syntax.md`.

File: kavm/src/tests/semantics/test_avm_scenarios.py [hunk 4]
```diff
@@ -25,13 +25,15 @@ def test_run_simulation(filename: str) -> None:
...
     avm_json_parser = project_path / kavm_definition_dir / 'parser_JSON_AVM-TESTING-SYNTAX'
+    teal_parser = project_path / kavm_definition_dir / 'parser_TealInputPgm_TEAL-PARSER-SYNTAX'
     teal_programs_parser = project_path / kavm_definition_dir / 'parser_TealProgramsStore_TEAL-PARSER-SYNTAX'
     proc_result = kavm.run_avm_json(
         input_file=Path(filename),
-        output='json',
+        output='none',
         profile=True,
         teal_sources_dir=Path(os.path.join(project_path, 'tests/teal-sources/')),
         teal_programs_parser=teal_programs_parser,
```
Comment (geo2a) [inline kavm/src/tests/semantics/test_avm_scenarios.py:35]:
```suggestion
```

After we have dropped this argument in the suggestion above.

File: kavm/src/kavm/__main__.py [hunk 5]
```diff
@@ -290,7 +295,13 @@ def create_argument_parser() -> ArgumentParser:
         '--teal-programs-parser',
         dest='teal_programs_parser',
         type=file_path,
-        help='Path to the executable to parse .teals files containing TealPrograms terms',
+        help='Path to the executable to parse TealProgramStore objects containing TealPrograms terms',
+    )
```
Comment (geo2a) [inline kavm/src/kavm/__main__.py:299]: The `--teal-programs-parser` option is obsolete now, since you're using the sort `Map[Id, TealInputPgm]`.

File: kavm/src/kavm/__main__.py [hunk 6]
```diff
@@ -101,6 +101,7 @@ def exec_run(
     input_file: Path,
     teal_sources_dir: Path,
     teal_programs_parser: Path,
```
Comment (geo2a) [inline kavm/src/kavm/__main__.py:103]:
```suggestion
```

File: kavm/src/kavm/__main__.py [hunk 7]
```diff
@@ -136,12 +139,14 @@ def exec_run(
                 print(proc_result.stdout)
             exit(proc_result.returncode)
         elif input_file.suffix == '.json':
+            print("test1")
             proc_result = kavm.run_avm_json(
                 input_file=input_file,
                 output=output,
                 profile=profile,
                 teal_sources_dir=teal_sources_dir,
                 teal_programs_parser=teal_programs_parser,
```
Comment (geo2a) [inline kavm/src/kavm/__main__.py:148]:
```suggestion
```

File: lib/include/kframework/avm/teal/teal-execution.md [hunk 8]
```diff
@@ -227,43 +227,55 @@ teal, failure means undoing changes made to the state (for more details, see
...
   // Leave the testing commands on the K cell
   rule <k> #stopIfError() ~> X:TestingCommand => X:TestingCommand ~> #stopIfError() ... </k>
 
   // Consume the rest of the K cell if the execution terminated with an error
-  rule <k> #stopIfError() ~> (_:KItem => .K) ... </k>
+  rule <k> #stopIfError() ~> (ITEM:KItem => .K) ... </k>
        <returncode> RETURN_CODE </returncode>
     requires RETURN_CODE =/=Int 0
-
+     andBool notBool(isTestingCommand(ITEM))
```
Comment (geo2a) [inline lib/include/kframework/avm/teal/teal-execution.md:278]: I don't think this condition is actually required. Let's add it for now though to be safe. We can test with `--coverage` later to see if it ever gets triggered.

File: lib/include/kframework/avm/teal/teal-syntax.md [hunk 9]
```diff
@@ -451,6 +451,7 @@ We define the syntax of TEAL's comments (using K's built-in sort `#Layout`), alo
   syntax Label           ::= r"({AlnumUbar}|{Special})+" [token]
   syntax HexToken        ::= r"0x{HexDigit}+"            [prec(2),token]
   syntax TAddressLiteral ::= r"[0-9A-Z]{58}"             [prec(1),token]
+  syntax #UpperID ::= TAddressLiteral [token]
```
Comment (geo2a) [inline lib/include/kframework/avm/teal/teal-syntax.md:454]: Why is this needed?

**Review by geo2a — CHANGES_REQUESTED** [review]
Great! I've requested some minor changes:
* Let's drop the `TealProgramsStore` sort, the `--teal-progams-parser` command-line option, and the `teal_programs_parser` argument to `kavm.run_avm_json` and `__main__.exec_run`.
* A couple of stray `print` functions need to be either dropped or refactored into logging calls with DEBUG level.

Also, I'm confused by the use of `#UpperID` in `teal-syntax.md`. Please elaborate.

Btw nice find about `kore-print`!

File: kavm/src/kavm/kavm.py (hunk 2 above)
Comment (nwatson22) [reply kavm/src/kavm/kavm.py]: removed as it was unintentionally left in.

File: kavm/src/kavm/kavm.py (hunk 3 above)
Comment (nwatson22) [reply kavm/src/kavm/kavm.py:204]: `TealProgramsStore` is now removed.

File: kavm/src/kavm/__main__.py (hunk 5 above)
Comment (nwatson22) [reply kavm/src/kavm/__main__.py]: I removed this option.

File: lib/include/kframework/avm/teal/teal-syntax.md (hunk 9 above)
Comment (nwatson22) [reply lib/include/kframework/avm/teal/teal-syntax.md]: I was running into a lexing issue with address literals which this seemed to fix. However, I can't remember the details, so I deleted it to see if it is still needed.

Comment (nwatson22) [reply lib/include/kframework/avm/teal/teal-syntax.md]: It seems to work fine without.

**Review by geo2a — APPROVED** [review]
